README.rst

```diff
@@ -1,102 +1,208 @@
+    (do io-monad
+      (putStrLn "hello")
+      (putStrLn "world")))
+
+You can run the above program using either stack or cabal::
+
+    $ cabal run klister -- run examples/hello.kl
+    hello
+    world
+
+Features
+========
+
+Features we borrow from Racket:
+
+* `Custom syntax`_, via `hygienic macros`_ with `easy-to-override hygiene`_.
+* `Custom languages`_ (``#lang``), via macros which reinterpret terms into
+  those of an existing ``#lang``.
+* `Syntax objects`_, that is, s-expressions annotated with source locations and
+  lexical information.
+* A `module system`_ which allow imports and exports to be renamed.
+
+.. _Custom syntax: examples/lambda-case.golden
+.. _hygienic macros: TODO: write a short example demonstrating lack of capture.
+.. _easy-to-override hygiene: examples/anaphoric-if.kl
+.. _Custom languages: examples/rpn.kl
+.. _Syntax objects: TODO: link to a short example which explains that in
+   Racket, syntax objects are introduced via ``#'(...)``, whereas in Klister
+   they are introduced via ``'(...)``. Also explain that Klister does not have
+   unannotated s-expressions.
+.. _module system: TODO: write a short example demonstrating what can be
+   expressed with the import and export primitives.
+
+Features we borrow from Haskell:
+
+* Statically-typed; we check that macro implementations have type
+  ``(-> Syntax (Macro Syntax)``, and *then* that the generated code is also
+  well-typed. We thus aim for the expressivity of Template Haskell, not the
+  extra guarantees of Typed Template Haskell.
+* Purely functional; primitives with side-effects either run in the ``IO``
+  monad or in the ``Macro`` monad, as appropriate.
+* Higher-kinded types; for example `monads`_ are defined as a library.
+* `Algebraic datatypes`_. In Racket, ``(list 1 2 3)`` and ``'(1 2 3)``
+  construct the same value, an s-expression which satisfies the ``list?``
+  predicate. In Klister, the former produces a value of the algebraic datatype
+  ``(List Integer)``, while the latter produces a value of type ``Syntax``.
+  This syntax can be "opened", via ``(open-syntax '(1 2 3))``, to obtain the
+  value ``(list-contents (list '1 '2 '3))`` of type
+  ``(Syntax-Contents Syntax)``.
+* `Bidirectional type inference`_; the type of an argument can specialize the
+  type of the function being called.
+* Bidirectional type inference; the type of a function can specialize the type
+  of its arguments.
+
+.. _monads: TODO: link to monad.kl's Monad definition, and add a comment there
+   highlighting the inferred type, especially the higher-kinded type variable.
+.. _Algebraic datatypes: TODO: write a small example defining and matching on
+   an algebraic type. Perhaps Either?
+.. _Bidirectional type inference: TODO: write a small example demonstrating
+   those two features.
+
+Features which make Klister special (but not *unique*; see the `bibliography`_):
+
+* `Type-providing macros`_; a macro used in an argument position can provide
```

Comment:
I think that this isn't about argument positions vs function positions. Instead, I'd say that we have interleaved type checking and macro expansion, so the results of macro expansion can provide type information to the rest of the program, and macros can query the expected type.

Reply:
Between these first two bullet points and the two "Bidirectional type inference" bullet points, I was trying to draw a parallel between, on one hand, the fact that unification allows type information to flow both in and out of a term, and on the other hand, the fact that our macros can both provide and receive type information. That's why I over-focus on "two directions", and needlessly mention the argument position in order to tie the bullet points together. I now see that I stretched the wording way too much and that the payoff isn't worth it, so I'll rewrite both pairs of bullet points.

Reply:
fixed in caaad94faa457e2204be892c7d4db8f5fc69620d